[SPARK-38582][K8S] Add `KubernetesUtils.buildEnvVars(WithFieldRef)?` utility functions

dcoliversun opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR introduces `buildEnvVars` and `buildEnvVarsWithFieldRef` functions for `kubernetesUtils` to eliminate duplicate code pattern.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Code simplification. Remove redundant EnvVar Build code for driver and executor pod.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
* Add unit test and Pass the GA.
* Running k8s integration test manually:
```plain
$ build/sbt -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube,r "kubernetes-integration-tests/test"

[info] KubernetesSuite:
[info] - Run SparkPi with no resources (18 seconds, 285 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (18 seconds, 56 milliseconds)
[info] - Run SparkPi with a very long application name. (15 seconds, 828 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (13 seconds, 807 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (14 seconds, 504 milliseconds)
[info] - Run SparkPi with an argument. (13 seconds, 573 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (17 seconds, 107 milliseconds)
[info] - All pods have the same service account by default (12 seconds, 551 milliseconds)
[info] - Run extraJVMOptions check on driver (7 seconds, 452 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (13 seconds, 910 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (26 seconds, 493 milliseconds)
[info] - Run SparkPi with env and mount secrets. (24 seconds, 546 milliseconds)
[info] - Run PySpark on simple pi.py example (14 seconds, 609 milliseconds)
[info] - Run PySpark to test a pyfiles example (17 seconds, 710 milliseconds)
[info] - Run PySpark with memory customization (14 seconds, 918 milliseconds)
[info] - Run in client mode. (10 seconds, 385 milliseconds)
[info] - Start pod creation from template (14 seconds, 522 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (12 seconds, 771 milliseconds)
[info] - Test basic decommissioning (48 seconds, 155 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (47 seconds, 723 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 48 seconds)
[info] - Test decommissioning timeouts (47 seconds, 866 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 45 seconds)
[info] Run completed in 13 minutes, 51 seconds.
[info] Total number of tests run: 23
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 23, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

dcoliversun committed:
[SPARK-38582][K8S] Introduce buildEnvVarsWithKV and buildEnvVarsWithFieldRef for KubernetesUtils to eliminate duplicate code pattern

dcoliversun commented:
@dongjoon-hyun 
It would be good if you could take a look when you have time, thanks!

Yikun reviewed (commented):
In general, I think refactoring helps to simplify the code. +1 on the idea. But for core featurestep changes we need to make sure test coverage all cases.

dcoliversun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,34 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
       }
     }
   }
+
```
Already have. Please check
https://github.com/apache/spark/blob/46ccc22ee40c780f6ae4a9af4562fb1ad10ccd9f/resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala#L51

https://github.com/apache/spark/blob/46ccc22ee40c780f6ae4a9af4562fb1ad10ccd9f/resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala#L506-L535

AmplabJenkins commented:
Can one of the admins verify this patch?

dcoliversun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+  /**
+   * This function builds the EnvVar objects for each key-value env.
+   */
+  @Since("3.3.1")
```
Nice. I will fix it.

dcoliversun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+  }
+
+  /**
+   * This function builds he EnvVar objects for each field ref env.
```
Nice. I will fix it.

dcoliversun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+   * This function builds the EnvVar objects for each key-value env.
+   */
+  @Since("3.3.1")
+  def buildEnvVarsWithKV(env: Seq[(String, String)]): Seq[EnvVar] = {
```
Good idea. `Map` is better than `Seq` because it is kv structure.

dcoliversun committed:
address comments

dcoliversun committed:
address comments

dcoliversun committed:
address comments

dcoliversun commented:
@Yikun @martin-g 
Update code for comments. It's good if you have a look again.

martin-g commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+   * This function builds the EnvVar objects for each key-value env.
+   */
+  @Since("3.4.0")
+  def buildEnvVarsWithKV(env: Map[String, String]): Seq[EnvVar] = {
```
```suggestion
  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
```
IMO there is no need of `WithKV` - the passed argument is not a Seq of tuples anymore.

dcoliversun replied:
Thanks for your comment :) Useful for me

martin-g commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,35 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
   }
+
+  test("SPARK-38582: verify that envVars built with kv env as expected") {
+    val input = for (i <- 9 to 1 by -1) yield (s"testEnvKey.$i", "testEnvValue123456")
```
```suggestion
    val input = for (i <- 9 to 1 by -1) yield (s"testEnvKey.$i", s"testEnvValue.$i")
```

dcoliversun replied:
Nice. I fixed it~

martin-g commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,35 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
+  }
+
+  test("SPARK-38582: verify that envVars built with field ref env as expected") {
+    val input = for (i <- 9 to 1 by -1) yield (s"testEnvKey.$i", s"v$i", "testEnvValue123456")
```
```suggestion
    val input = for (i <- 9 to 1 by -1) yield (s"testEnvKey.$i", s"v$i", s"testEnvValue.$i")
```

dcoliversun replied:
Nice. I fixed it~

martin-g reviewed (approved):
LGTM!
The suggestions below are just nits which you could ignore!

dcoliversun committed:
address comments

dcoliversun commented:
Gentle ping @dongjoon-hyun @martin-g

martin-g commented:
I am just a contributor to the project. A project member could merge PRs.
// CC @dongjoon-hyun @HyukjinKwon @holdenk

dongjoon-hyun commented:
Sorry for being late. To prepare Apache Spark 3.3.0, I was too busy in these days.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
+    if (env.isEmpty) {
```
I'm wondering why we need this. This doesn't look like `Scalatic` if `else` is enough to handle this case.

dcoliversun replied:
`else` was removed in https://github.com/apache/spark/pull/35886/commits/cfb25897c87d7fef0ead07067b7194d60989851f

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,42 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
+    if (env.isEmpty) {
```
ditto. This looks like redundant and not `Scalatic`.

dcoliversun replied:
`if` was removed in https://github.com/apache/spark/pull/35886/commits/cfb25897c87d7fef0ead07067b7194d60989851f

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala:
```diff
@@ -74,16 +74,8 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
-      }
-
+    val driverCustomEnvs = KubernetesUtils.buildEnvVars(
+      conf.environment + (ENV_APPLICATION_ID -> conf.appId))
```
Ur, are you sure about this change? This looks inconsistent to me because this breaks the existing behavior. Previously, `conf.environment`'s `ENV_APPLICATION_ID` supersedes if there is a duplication.

dcoliversun replied:
Your worries are right. After the latest commit, I don't think this refactoring will introduce lots of regressions.
1. When there are two environment variables with the same key in the pod, the last value shall work.

I apply a pod with yaml
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: static-web
  labels:
    role: myrole
spec:
  containers:
    - name: web
      image: nginx
      env:
      - name: DEMO_GREETING
        value: "Hello from the environment"
      - name: DEMO_GREETING
        value: "Such a sweet sorrow"
      ports:
        - name: web
          containerPort: 80
          protocol: TCP
```
And execute the echo command, get the value of `DEMO_GREETING` is `Such a sweet sorrow`.
```shell
$ kubectl exec -ti static-web -n default -- /bin/bash
$ echo $DEMO_GREETING
Such a sweet sorrow
```
2. `++` operation in Scala Map is overwrite append operation. As long as the order of our `++` remains the same as the previous `env` order, the value of the environment variable in the pod will not change. For example, user specified `ENV_APPLICATION_ID` as `temp_app_id` in sparkConf
```plain
spark.kubernetes.driverEnv.ENV_APPLICATION_ID=user_specified_app_id
```
According to the previous logic, two envs with the same key will be generated
```yaml
env:
  - name: ENV_APPLICATION_ID
    value: spark-XXXXXXX # https://github.com/apache/spark/blob/b852645f69b3b7a0a2140a732c4c03b302f8795a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala#L222
  - name: ENV_APPLICATION_ID
     value: user_specified_app_id
```
According to the first point, the value of `ENV_APPLICATION_ID` is `user_specified_app_id`.

We replace `seq` with `map`. It causes the env as follow
```yaml
env:
  - name: ENV_APPLICATION_ID
     value: user_specified_app_id
```
Because `++` in `map` is overwrite operation, so the value of `ENV_APPLICATION_ID` is overwritten. Finally, the value of `ENV_APPLICATION_ID` in pod is `user_specified_app_id`.

I agree that this refactoring does not bring env changes and can reduce unnecessary env.

dongjoon-hyun reviewed (changes requested):
+1 for the intention, but the proposed PR looks a little intrusive and inconsistent to me. 

This PR may introduce lots of regression if we are not careful enough to avoid any regression like the following comment.
- https://github.com/apache/spark/pull/35886/files#r841159948

dongjoon-hyun commented:
FYI, we had better hold on these kind of PRs during the planned release process. It's the same for the other refactoring PRs.
- https://github.com/apache/spark/pull/36049#pullrequestreview-929704033

dcoliversun commented:
@dongjoon-hyun Hi, I'm ok to hold on this PR during the planned release process. Thanks for your work on the release plan. And I'm addressing your comment.

dongjoon-hyun commented:
Thank you, @dcoliversun !

dcoliversun committed:
address comments

dcoliversun committed:
address comments

dcoliversun committed:
Merge branch 'master' into SPARK-38582

dcoliversun committed:
Merge branch 'master' into SPARK-38582

dcoliversun commented:
> +1 for the intention, but the proposed PR looks a little intrusive and inconsistent to me.
> 
> This PR may introduce lots of regression if we are not careful enough to avoid any regression like the following comment.
> 
> * https://github.com/apache/spark/pull/35886/files#r841159948

@dongjoon-hyun Hi, reply about this worry is here 
https://github.com/apache/spark/pull/35886#discussion_r843447694

dongjoon-hyun commented:
Thank you for your reply, @dcoliversun .

dongjoon-hyun commented:
Just an update for you, @dcoliversun . 

Apache Spark community is still focusing on 3.3.0 release. (RC2 vote seems to fail and RC3 vote will start.)
This PR is not under full review yet because of the nature of intrusive refactoring patterns.
We will revisit this after 3.3.0 release. At that time, we will discuss about the tradeoff between the increased maintenance cost due to this PR and the benefit.

Thank you for your patience.

dongjoon-hyun reviewed (commented):
Hi, @dcoliversun . I'm revisiting the pending K8s PRs for Apache Spark 3.4.0. Could you resolve the conflicts?

dcoliversun commented:
Thank you @dongjoon-hyun  I'd like to do this.

dcoliversun committed:
Merge branch 'master' into SPARK-38582

dcoliversun commented:
Conflict caused by https://github.com/apache/spark/pull/36716
```scala
// before [SPARK-39062](https://issues.apache.org/jira/browse/SPARK-39062)
(ENV_EXECUTOR_CORES, execResources.cores.toString),
// after [SPARK-39062](https://issues.apache.org/jira/browse/SPARK-39062)
(ENV_EXECUTOR_CORES, execResources.cores.get.toString),
```

dcoliversun committed:
add comma before ENV_EXECUTOR_MEMORY

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
+      env.filter( env => env._2 != null)
```
Apache Spark uses 2-space indentation. Could you adjust this?

dongjoon-hyun replied:
Also, `( env =>` -> `(env =>`

dcoliversun replied:
Yes, updated

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
+      env.filter( env => env._2 != null && env._3 != null)
```
ditto. Two-space indentation and removal of space by `( end =>` -> `(end =>`.

dcoliversun replied:
updated

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala:
```diff
@@ -84,25 +84,18 @@ private[spark] class DriverCommandFeatureStep(conf: KubernetesDriverConf)
+            .orElse(environmentVariables.get(ENV_PYSPARK_DRIVER_PYTHON))
+            .orElse(environmentVariables.get(ENV_PYSPARK_PYTHON))
+            .orNull))
+    }
```
Could you check what happens when both `PYSPARK_PYTHON` and `ENV_PYSPARK_PYTHON` is not defined? New logic looks not identical with the previous one.

dcoliversun replied:
Let me check it

dcoliversun replied:
Yes, before [d2395f0](https://github.com/apache/spark/pull/35886/commits/d2395f0ea33151243c43faebb2894a7585608d3b), new logic is not identical.  
### Old logic
https://github.com/apache/spark/blob/36dd531a93af55ce5c2bfd8d275814ccb2846962/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala#L87-L105
In old logic, when both `PYSPARK_PYTHON` and `ENV_PYSPARK_PYTHON` is not defined, spark pod has empty environment variables, which of keys are `ENV_PYSPARK_PYTHON` and `ENV_PYSPARK_DRIVER_PYTHON`, and values are `NULL`.  The corresponding Pod yaml is as follows  
```yaml
...
spec:
  ...
  env:
  - name: PYSPARK_PYTHON
  - name: PYSPARK_DRIVER_PYTHON
```
Exec `echo` command, the result is as follows. 
```shell
$ echo $PYSPARK_PYTHON

$ echo $PYSPARK_DRIVER_PYTHON

```
### New logic
```scala
  val pythonEnvs = {
      KubernetesUtils.buildEnvVars(
        Map(
          ENV_PYSPARK_PYTHON -> conf.get(PYSPARK_PYTHON)
            .orElse(environmentVariables.get(ENV_PYSPARK_PYTHON))
            .orNull,
          ENV_PYSPARK_DRIVER_PYTHON -> conf.get(PYSPARK_DRIVER_PYTHON)
            .orElse(conf.get(PYSPARK_PYTHON))
            .orElse(environmentVariables.get(ENV_PYSPARK_DRIVER_PYTHON))
            .orElse(environmentVariables.get(ENV_PYSPARK_PYTHON))
            .orNull))
    }

  /**
   * This function builds the EnvVar objects for each key-value env.
   */
  @Since("3.4.0")
  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
    env.map { env =>
      new EnvVarBuilder()
        .withName(env._1)
        .withValue(env._2)
        .build()
    }.toSeq
  }

  /**
   * This function builds the EnvVar objects for each field ref env.
   */
  @Since("3.4.0")
  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
    env.map { env =>
      new EnvVarBuilder()
        .withName(env._1)
        .withValueFrom(new EnvVarSourceBuilder()
          .withNewFieldRef(env._2, env._3)
          .build())
        .build()
    }
  }
```
Now, [d2395f0](https://github.com/apache/spark/pull/35886/commits/d2395f0ea33151243c43faebb2894a7585608d3b) allows to set empty environment variables on pod. When second value of `env` is null in `buildEnvVars` method, it still used in `EnvVarBuilder method`.
```scala
> print(env)
Some(EnvVar(name=PYSPARK_DRIVER_PYTHON, value=null, valueFrom=null, additionalProperties={}))
```
I believe [d2395f0](https://github.com/apache/spark/pull/35886/commits/d2395f0ea33151243c43faebb2894a7585608d3b) can keep identical with the previous one.  
Unit test about case that both `PYSPARK_PYTHON` and `ENV_PYSPARK_PYTHON` is not defined is added.

dongjoon-hyun replied:
Thank you for in-depth analysis.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:
```diff
@@ -122,68 +122,41 @@ private[spark] class BasicExecutorFeatureStep(
+        Seq(
+          (ENV_EXECUTOR_POD_IP, "v1", "status.podIP"),
+          (ENV_EXECUTOR_POD_NAME, "v1", "metadata.name")
+        ))
```
Technically, new logic looks inconsistent to me in terms of the order of variable definition.
New logic seems to have a limitation where `buildEnvVars`-defined ones are overwritten by `buildEnvVarsWithFieldRef`-defined ones always. So, do we assume variable uniqueness here?

dcoliversun replied:
I think we need variable uniqueness. As you said, `envVar` is overwrite in `pod.spec.container.envVar`.
* `buildEnvVarsWithFieldRef` method can project Pod-level fields into the running container as environment variables, which value is defined until pod scheduled or later.
* `buildEnvVars` method only put prepared key-value into pre-submission container as environment variables.  
From normal usage, I believe `buildEnvVarsWithFieldRef` has higher priority than `buildEnvVars`.

dongjoon-hyun replied:
Ack, @dcoliversun .

dcoliversun committed:
fix scala code style

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:
```diff
@@ -122,68 +122,41 @@ private[spark] class BasicExecutorFeatureStep(
+        Map(SecurityManager.ENV_AUTH_SECRET -> secMgr.getSecretKey())
+      } else {
+        Nil
+      }
```
To be consistent, we need to skip `Map` creation when `secMgr.getSecretKey()` is null too.
https://github.com/apache/spark/blob/5bffb983f8f9482660701cc34e7cc23771961888/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala#L156-L161

dcoliversun replied:
Yes,`Option().map` looks like more `Scalatic`

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,35 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
     }
   }
+
+  test("SPARK-38582: verify that envVars built with kv env as expected") {
```
`built` -> `is built`?

dcoliversun replied:
My grammatical mistake. Thanks.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,35 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
+    assert(outputEnvVars.toSet == expectedEnvVars.toSet)
+  }
+
+  test("SPARK-38582: verify that envVars built with field ref env as expected") {
```
`built` -> `is built`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:
```diff
@@ -122,68 +122,41 @@ private[spark] class BasicExecutorFeatureStep(
+          ENV_RESOURCE_PROFILE_ID -> resourceProfile.id.toString)
+          ++ kubernetesConf.environment
+          ++ sparkAuthSecret
+          ++ Map(ENV_CLASSPATH-> kubernetesConf.get(EXECUTOR_CLASS_PATH).orNull)
```
Indentation? `ENV_CLASSPATH->` -> `ENV_CLASSPATH ->`

dcoliversun replied:
oh, my mistake. Thanks, updated in [41f27b1](https://github.com/apache/spark/pull/35886/commits/41f27b16f5185571ffd1e585eeb11c7815f00dfb)

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
   }
+
+  /**
+   * This function builds the EnvVar objects for each key-value env.
```
The function description is inconsistent from the code body because ew don't build EnvVar when value is null.

dcoliversun replied:
I think we should mention it, because it's supported to set on yaml and works normally in container.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+  }
+
+  /**
+   * This function builds the EnvVar objects for each field ref env.
```
ditto. This function description is inconsistent from the code body because we don't build it when the second and the third values are null.

dcoliversun replied:
Thanks. I choose `no-null apiVersion and fieldPath` here

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesUtilsSuite.scala:
```diff
@@ -127,4 +127,35 @@ class KubernetesUtilsSuite extends SparkFunSuite with PrivateMethodTester {
+      new EnvVarBuilder()
+        .withName(k)
+        .withValue(v).build()
+    }
```
Please add a case with an empty string value, `''`.

dcoliversun replied:
Added and key-only envVar is expected.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
+    env.filter(env => env._2 != null)
```
```scala
-    env.filter(env => env._2 != null)
+    env.filter(_._2 != null)
```

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+  @Since("3.4.0")
+  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
+    env.filter(env => env._2 != null)
+      .map { env =>
```
ditto. Please avoid to use `env` in this context.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   * This function builds the EnvVar objects for each key-value env.
+   */
+  @Since("3.4.0")
+  def buildEnvVars(env: Map[String, String]): Seq[EnvVar] = {
```
Question. Why do we need `Map` here? If we can, shall we use `Seq` instead of `Map` simply?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   * This function builds the EnvVar objects for each field ref env.
+   */
+  @Since("3.4.0")
+  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
```
Unlike `buildEnvVars`, in this method, we already use `Seq`.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+   */
+  @Since("3.4.0")
+  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
+    env.filter(env => env._2 != null && env._3 != null)
```
Let's avoid to use the same name, `env` for item with the above collection, `env`.

dcoliversun replied:
Updated.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -381,4 +381,34 @@ object KubernetesUtils extends Logging {
+  @Since("3.4.0")
+  def buildEnvVarsWithFieldRef(env: Seq[(String, String, String)]): Seq[EnvVar] = {
+    env.filter(env => env._2 != null && env._3 != null)
+      .map { env =>
```
ditto.

dcoliversun replied:
Replaced by `case (key, apiVersion, fieldPath)`. I believe that it's more meaningful.

dongjoon-hyun committed:
Use Seq instead of Map

dongjoon-hyun reviewed (changes requested):
I made a PR to you, @dcoliversun .
- https://github.com/dcoliversun/spark/pull/4

dcoliversun committed:
Merge pull request #4 from dongjoon-hyun/use_seq

Use Seq instead of Map

dcoliversun commented:
@dongjoon-hyun Thanks for you PR and advice :) I'm working on it. Update later

dcoliversun committed:
address comments

address comments

address comments

address comments

dcoliversun committed:
fix ut error

dongjoon-hyun committed:
Fix typos

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @dcoliversun . 
I fixed a few typos at the last commit.
Merged to master for Apache Spark 3.4.0.

dongjoon-hyun commented:
Also, thank you, @Yikun and @martin-g , too.

dcoliversun commented:
Thanks for your help @dongjoon-hyun
